src/main/groovy/com/netflix/nebula/lint/plugin/GradleLintPlugin.groovy

```diff
@@ -37,47 +39,47 @@ class GradleLintPlugin implements Plugin<Project> {
             manualLintTask.group = 'lint'
             manualLintTask.failOnWarning = true
 
+            def criticalLintTask = project.tasks.create('criticalLintGradle', LintGradleTask)
+            criticalLintTask.group = 'lint'
+            criticalLintTask.onlyCriticalRules = true
+
             def fixTask = project.tasks.create('fixGradleLint', FixGradleLintTask)
             fixTask.userDefinedListeners = lintExt.listeners
 
             def fixTask2 = project.tasks.create('fixLintGradle', FixGradleLintTask)
             fixTask2.userDefinedListeners = lintExt.listeners
 
-            autoLintTask.onlyIf {
-                def allTasks = project.gradle.taskGraph.allTasks
-                def hasFailedTask = !lintExt.autoLintAfterFailure && allTasks.any { it.state.failure != null }
-                def hasExplicitLintTask = allTasks.any { it == fixTask || it == fixTask2 || it == manualLintTask }
-                !hasFailedTask && !hasExplicitLintTask
-            }
-        }
+            project.gradle.addBuildListener(new LintListener() {
+                def allTasks
 
-        configureReportTask(project, lintExt)
+                @Override
+                void graphPopulated(TaskExecutionGraph graph) {
+                    allTasks = graph.allTasks
+                }
 
-        def finalizeByLint = { task ->
-            if(lintExt.alwaysRun) {
-                def rootLint = project.rootProject.tasks.getByName('autoLintGradle')
-                if (task != rootLint && !exemptTasks.contains(task.name)) {
-                    // when running a lint-eligible task on a subproject, we want to lint the whole project
-                    task.finalizedBy rootLint
-
-                    // because technically you can override path in a Gradle task implementation and cause path to be null!
-                    if (task.getPath() != null) {
-                        try {
-                            rootLint.shouldRunAfter task
-                        } catch (Throwable ignored) {
-                            // just quietly DON'T add rootLint to run after this task, it will probably still run because
-                            // it will be hung on some other task as a shouldRunAfter
-                        }
+                @Override
+                void buildFinished(BuildResult result) {
+                    if (onlyIf()) {
+                        autoLintTask.lint()
```

Comment:
What do you think about using a task class directly like this?

Reply:
it works, do you know if you can add a task to the taskGraph when in `project.gradle.taskGraph.whenReady`

if the taskGraph thing doesn't work, i'm good with buildFinished